Add bft explanations and guide - Closes #743 (#750)
LiskHQ/lisk-docs: modules/ROOT/pages/explanations/consensus.adoc

:url_protocol_punishment: {v_protocol}@lisk-protocol::consensus-algorithm.adoc#punishment
:url_protocol_transactions_pom: {v_protocol}@lisk-protocol::transactions.adoc#pom
:url_protocol_transactions_delegate: {v_protocol}@lisk-protocol::transactions.adoc#delegate
:url_protocol_transactions_unlock: {v_protocol}@lisk-protocol::transactions.adoc#unlock
:url_protocol_transactions_vote: {v_protocol}@lisk-protocol::transactions.adoc#vote

[[dpos]]
== Delegated Proof of Stake (DPoS)

The Lisk SDK bootstraps a blockchain network that is based on the xref:{url_protocol_dpos}[Lisk DPoS] consensus algorithm.

In DPoS based blockchains, the consensus regarding who can forge the next block is reached by users according to the votes cast.

[NOTE]
====
The DPoS used by Lisk is in fact more of a middle ground between {url_wiki_pos}[PoS] and DPoS.

The DPoS related characteristic is the ability of users to <<delegate_registration,register as delegate>> and then <<vote,receive votes>> from other users, in order to increase their <<weight,delegate weight>>.

The PoS related characteristic is the requirement for delegates to self-vote a certain amount of tokens, to increase their delegate weight.

> Another PoS characteristic is the mechanism for the selection of the <<forging_delegates,two random standby delegates>>, who are selected each forging round.

Reviewer: ```suggestion
Another PoS characteristic is the mechanism for the selection of the <<forging_delegates,two random standby delegates>>, who are both selected for a forging round.
```

Author: ```suggestion
Another PoS characteristic is the mechanism for the selection of the <<forging_delegates,two random standby delegates>>, who are selected every forging round.
```

> Here applies, the higher the delegate weight, the higher the chance to be selected in one of the two random spots for standby delegates.

Reviewer: ```suggestion
Hence, the higher the delegate weight, the higher the chance to be selected in one of the two random spots for standby delegates.
```

Author: ```suggestion
The higher the delegate weight, the higher the chance to be selected in one of the two random spots for standby delegates.
```

:url_guides_forging_safely: guides/node-management/forging.adoc#safely_activating_forging

:url_protocol_blocks_rewards: {v_protocol}@lisk-protocol::blocks.adoc#rewards_and_fees
:url_protocol_dpos: {v_protocol}@lisk-protocol::consensus-algorithm.adoc#dpos
:url_protocol_bft: {v_protocol}@lisk-protocol::consensus-algorithm.adoc#lisk_bft
:url_protocol_delegate_selection: {v_protocol}@lisk-protocol::consensus-algorithm.adoc#delegate_selection
:url_protocol_fork_choice: {v_protocol}@lisk-protocol::consensus-algorithm.adoc#index-fork_choice_rule-1
:url_protocol_punishment: {v_protocol}@lisk-protocol::consensus-algorithm.adoc#punishment
:url_protocol_transactions_pom: {v_protocol}@lisk-protocol::transactions.adoc#pom
:url_protocol_transactions_delegate: {v_protocol}@lisk-protocol::transactions.adoc#delegate
:url_protocol_transactions_unlock: {v_protocol}@lisk-protocol::transactions.adoc#unlock
:url_protocol_transactions_vote: {v_protocol}@lisk-protocol::transactions.adoc#vote

[[dpos]]
== Delegated Proof of Stake (DPoS)

The Lisk SDK bootstraps a blockchain network that is based on the xref:{url_protocol_dpos}[Lisk DPoS] consensus algorithm.

> In DPoS based blockchains, the consensus regarding who can forge the next block is reached by users who vote with their stake for other users.

Reviewer 2: ```suggestion
In DPoS based blockchains, the consensus regarding who can forge the next block is reached by users according to the votes cast. 
```

====

[[forging_delegates]]
....
101(amount of the active delegates) + 2(random standby delegates) = 103(Number of blocks of a forging round)
....

More information about the delegate selection mechanism can be found in the xref:{url_protocol_delegate_selection}[Lisk Protocol].

[[weight]]
==== Delegate weight

The 101 delegates with the highest delegate weight are selected for the active forging positions.

The delegate weight is defined as shown below:

....
delegate weight = minimum { 10 * delegate self-vote , sum of all votes for the delegate }
....

Where `delegate self-vote` is the amount the delegate voted for its own account.

The `sum of all votes for the delegate` includes the self-votes.

[[delegate_registration]]
=== Registering as delegate

In a DPoS system, each account that has an adequate enough balance to send a xref:{url_protocol_transactions_delegate}[register delegate transaction] can register a new delegate on the network.
Other accounts can <<vote, vote for delegates>> to support them with their stake.

> As a reward for securing the network, the forging delegates receive the xref:{url_protocol_blocks_rewards}[transaction fees and block rewards] of the forged blocks and the transactions included int he forged blocks.

Reviewer: ```suggestion
As a reward for securing the network, the forging delegates receive the xref:{url_protocol_blocks_rewards}[transaction fees and block rewards] of the forged blocks, and the transactions included in the forged blocks.
```

[[vote]]
=== Voting for a delegate

Accounts can vote for delegates by sending a xref:{url_protocol_transactions_vote}[delegate vote transaction].

> The locked tokens can be unlocked again if needed, by <<unvote, unvoting>> the delegate again.

Reviewer 2: I think here we are missing the info that "voting will lock your tokens".

Reviewer 2: This info comes later (in the next section). Maybe it makes sense to have it here (as the line starts with "The locked tokens ...")

The amount of tokens used for voting is locked and cannot be used for any other transactions.
This includes but is not limited to balance transfers, further voting or fees.

To use those tokens, the account has to submit a xref:{url_protocol_transactions_vote}[delegate vote transaction], with a *negative amount* (also called “unvote”).

> This will start the unlocking procedure and the LSK will be ready for unlocking 2000 blocks later (roughly 5 hours and 30 minutes).

Reviewer: ```suggestion
This will start the unlocking procedure and the LSK will be ready for unlocking 2000 blocks later, (approximately 5 hours and 30 minutes).
```

});
----

.An unlock transaction can contain multiple unlock objects
[TIP]
====
This allows an account to submit multiple delegate votes and recover those tokens with a single unlock transaction.
Of course, all tokens must have been in the unlocking state for at least 2000 blocks for the unlock to be valid.
====

== BFT

The Byzantine Fault Tolerance (BFT) algorithm ensures that the network can always reach a consensus about the current state of the blockchain.
The main purpose of the BFT consensus algorithm is that for a given height eventually all Lisk nodes agree on the same block.
This is in particular important in the case where there are different valid blocks for the same height, which can occur due to network delays or delegates forging multiple blocks in their designated time slot.

> In most cases, the BFT consensus algorithm implemented in the Lisk SDK takes care of these matters and resolves all conflicts automatically, following rules defined in the xref:{url_protocol_bft}[BFT] section of the Lisk Protocol.

Reviewer: ```suggestion
In most cases, the BFT consensus algorithm implemented in the Lisk SDK takes care of these points and resolves all conflicts automatically, following the rules defined in the xref:{url_protocol_bft}[BFT] section of the Lisk Protocol.
```

Additionally delegates can be punished by anyone in the network, if they forge a block with consensus votes that contradict with the Lisk BFT (see the xref:{url_protocol_punishment}[Punishment of Lisk-BFT protocol violations] section of the Lisk Protocol).
This can happen for example if they...

> * ... perform double forging, i.e. having forging actived on multiple node for the same delegate at the same time.

Reviewer: ```suggestion
* ... Double forging, i.e. having forging activated on multiple nodes for the same delegate at the same time.
```

> * ... violate the xref:{url_protocol_fork_choice}[fork choice rule].

Reviewer: ```suggestion
* ... Violating the xref:{url_protocol_fork_choice}[fork choice rule].
```

> * ... xref:{url_guides_forging_safely}[enable forging on a new node], and didn't port over the `forger_info` data.

Reviewer: ```suggestion
* ... xref:{url_guides_forging_safely}[Enabling forging on a new node], without porting over the `forger_info` data.
```

=== Delegate misbehaviour report

A  xref:{url_protocol_transactions_pom}[delegate misbehaviour report] transaction can be issued by anyone in the network, who observes a violation of the Lisk BFT consensus algorithm.

> A misbehaviour of a delegate is indicated by the `maxHeightPreviouslyForged` property of a block forged by a delegate.

Reviewer: ```suggestion
Any misbehavior of a delegate is indicated by the `maxHeightPreviouslyForged` property of a block forged by a delegate.
```

> A misbehaviour of a delegate is indicated by the `maxHeightPreviouslyForged` property of a block forged by a delegate.

Reviewer 2: A misbehavior can also occur by violation of the fork choice rule. Technically, a violation is indicated by 2 contradicting block headers.
The definition of contradicting is given in the BFT LIP (https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#detecting-contradicting-block-headers). 

It boils down to 3 different cases, one of them concerning the `maxHeightPreviouslyForged`.

Author: Thanks for the link! I added now a new section with that snippet from the LIP, it's quite helpful :)

> If `maxHeightPreviouslyForged` is not equal to the `height` of the block which was previousely forged by that delegate, this can be reported by providing both block headers of the delegate, which have contradicting values.

Reviewer: ```suggestion
If `maxHeightPreviouslyForged` is not equal to the `height` of the block which was previously forged by the delegate, this can be reported by providing both block headers of the delegate, which have contradicting values.
```

> If `maxHeightPreviouslyForged` is not equal to the `height` of the block which was previousely forged by that delegate, this can be reported by providing both block headers of the delegate, which have contradicting values.

Reviewer 2: This is slightly incorrect. The `maxHeightPreviouslyForged` can be different from the last height forged on the chain by the delegate, and this might not be a misbehavior.

Reviewer 2: A delegate misbehavior is indicated by 2 contradicting block headers signed by the delegate. The definition of contradicting is given in the BFT LIP (https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#detecting-contradicting-block-headers). 

It boils down to 3 different cases, one of them concerning the `maxHeightPreviouslyForged`.

            previousBlockId: '9326981395427095175',
            reward: '500000000',
            seedReveal: 'abe2a66d7a35fd7b580e977d9f7911ae',
            timestamp: 122329567,
            totalAmount: '0',
            totalFee: '0',
            version: 2
        },
        header2: {
            blockSignature: '31ccf4ce1a3a224a2a32c3f4bdc6fad0ddb8feb45b05b7d411eee1a608f9d91284d09c727bba173c882d5dc90cb951c5affc10462d650031a627e00d919cbf08',
            generatorPublicKey: '7a7f24c061db6a92320ba14323f814c20dbcc811a931ead3ca63c75a4de1b643',
            height: 8933,
            maxHeightPreviouslyForged: 8788,
            maxHeightPrevoted: 8868,
            numberOfTransactions: 0,
            payloadHash: 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
            payloadLength: 0,
            previousBlockId: '9326981395427095175',
            reward: '500000000',
            seedReveal: 'abe2a66d7a35fd7b580e977d9f7911ae',
            timestamp: 122329567,
            totalAmount: '0',
            totalFee: '0',
            version: 2
        }
    }
});
----

[[punishment_consequences]]
=== Consequences for punished delegates and their voters

> If a valid delegate misbehaviour report is posted to the network, the respective delegate will face the following consequences:

Reviewer: ```suggestion
If a valid delegate misbehavior report is posted to the network, the respective delegate will face the following consequences:
```

> * setting the delegate weight to 0 for the next 780,000 blocks (approximately 3 month).

Reviewer: ```suggestion
* Setting the delegate weight to 0 for the next 780,000 blocks (approximately 3 months).
```

> * the unlocking period for self-votes is increased from 260,000 blocks to 780,000 blocks (from approximately 1 month to 3 months) for the respective delegate.

Reviewer: ```suggestion
* The unlocking period for self-votes is increased from 260,000 blocks to 780,000 blocks (from approximately 1 month to 3 months), for the respective delegate.
```

Voters of the respective delegate will receive the following punishment:

> * the unlocking period for the votes for the punished delegate is increased from 2,000 blocks to 260,000 blocks (from approximately 5 hours to 1 month).

Reviewer: ```suggestion
* The unlocking period for the votes for the punished delegate is increased from 2,000 blocks to 260,000 blocks (from approximately 5 hours to 1 month).
```

});
----

.An unlock transaction can contain multiple unlock objects
[TIP]
====
This allows an account to submit multiple delegate votes and recover those tokens with a single unlock transaction.
Of course, all tokens must have been in the unlocking state for at least 2000 blocks for the unlock to be valid.
====

== BFT

> The Byzantine Fault Tolerance (BFT) algorithm ensures that the network can always reach a consensus about the current state of the blockchain.

Reviewer 3: ```suggestion
The Byzantine Fault Tolerance (BFT) algorithm ensures that the network can reach consensus about the current state of the blockchain.
```

> The main purpose of the BFT consensus algorithm is to ensure that for a given height, eventually all Lisk nodes agree on the same block.

Reviewer 3: ```suggestion
This means that for a given height, eventually all Lisk nodes agree on the same block.
```

This is particularly important in the case where there are different valid blocks for the same height, which can occur due to network delays or delegates forging multiple blocks in their designated time slot.

> Additionally, delegates can be punished by anyone in the network if they forge contradicting blocks, i.e., two blocks with consensus votes that violate the Lisk-BFT protocol (see the xref:{url_protocol_punishment}[Lisk protocol documentation] for details).

Reviewer 3: The link is broken.

Author: For me it works fine 🤔

Reviewer 3: Okay. Maybe it is just not working on GitHub.

> In most cases, the BFT consensus algorithm implemented in the Lisk SDK takes care of these points and resolves all conflicts automatically, following the rules defined in the xref:{url_protocol_bft}[BFT] section of the Lisk Protocol.

Reviewer 3: I am not sure what we are trying to say here. What are these points and what are the conflicts (you mean conflicting branches/blocks?). Do we try to emphasize that no node operator intervention is necessary for the consensus algorithm?

Author: Because we introduce in the next paragraph the delegate misbehavior report, where users have to do something manually. This sentence says it has also ways of automatically reaching consensus in the network. As far as I understand, the fork choice rule is ensuring this. In the past we had a lot of more fork scenarios than we have now with the new BFT, so I assume the fork choice rule is taking care of these situations now automatically?

If so, we could maybe change it like below?

```suggestion
In most cases, the BFT consensus algorithm implemented in the Lisk SDK takes care of these points and resolves all forks automatically, following the fork choice rule defined in the xref:{url_protocol_bft}[BFT] section of the Lisk Protocol.
```

Reviewer 3: Then I would rather remove this sentence. We already explain before what consensus is and that Lisk-BFT ensures that the network reaches consensus. Otherwise, we would need to introduce a bit more terminology (fork choice rule, finality,...).

Author: Ok, I removed the sentence now!

> Additionally delegates can be punished by anyone in the network, if they forge a block with consensus votes that contradict with the Lisk BFT (see the xref:{url_protocol_punishment}[Punishment of Lisk-BFT protocol violations] section of the Lisk Protocol).

Reviewer 3: ```suggestion
Additionally, delegates can be punished by anyone in the network, if they forge contradicting blocks, i.e., two blocks with consensus votes that violate the Lisk BFT (see the xref:{url_protocol_punishment}[Punishment of Lisk-BFT protocol violations] section of the Lisk Protocol).
```

Reviewer 3: I would maybe shorten the reference to "see the protocol documentation for details" and still link directly to the specific section.

> This can happen for example if the following actions are performed:

Reviewer 3: Maybe we could phrase it positively, i.e., say what a node operator has to do to ensure that they are never punished. For example:

> In order to avoid that a delegate is punished, a node operator has to take a lot of care when enabling forging on a node. This means that they have to ensure the following:
* Never activate forging on more than one node. Otherwise, the node operator risks to double-forge, i.e., produce two contradicting blocks for the same time slot which can lead to punishment.
* Never activate forging without porting over the `forger_info` data. Otherwise, the `lastHeightPreviouslyForged` property of the forged blocks may be outdated, which can cause the delegate to forge contradicting blocks.

* ... Double forging, i.e. having forging activated on multiple nodes for the same delegate at the same time.

> * ... Violating the xref:{url_protocol_fork_choice}[fork choice rule].

Reviewer 3: This should not happen if you don't modify the BFT related code. So maybe I would remove it from the list. Then the list only contains things that the node operator could do wrong which cause a punishment of the delegate.

* ... xref:{url_guides_forging_safely}[Enabling forging on a new node], without porting over the `forger_info` data.

=== Delegate misbehaviour report

[[discover]]
=== How to discover misbehavior of a delegate

> Any misbehavior of a delegate is indicated by contradicting block headers between two forged blocks of a particular delegate.

Reviewer 3: ```suggestion
Any misbehavior of a delegate is indicated by two contradicting block headers signed by the delegate.
```

> This can be reported by providing the two block headers which have contradicting values.

Reviewer 3: ```suggestion
A misbehavior can therefore be reported by providing the two contradicting block headers.
```

Reviewer 3: Contradicting values maybe understood as just different values.

Reviewer 3: I am not 100 % if I would put the code here as it requires quite a good understanding of Lisk-BFT to fully understand it.

Author: Imho the snippet describes the process of finding contradicting blocks pretty nice in pseudo code.
The comments in the code also help with understanding it.

> .Detecting Contradicting Block Headers (Source: {url_github_lip14}[LIP 14^])

Reviewer 3: Maybe I would rather link to the SDK code here as some variable names could be slightly different.

Author: Is this implemented in SDK v4 already? I thought it's coming later

Reviewer 3: The code is here https://github.com/LiskHQ/lisk-sdk/blob/00917106c9d6bfd88fdceb6d59aad01b80ce021f/elements/lisk-bft/src/finality_manager.ts#L368
The code exists because you cannot add two contradicting blocks to the same branch. But maybe it is not as easy to understand.

      b1=blockHeader2;
      b2=blockHeader1;
   }

   // The order of cases is essential here
   if(b1.delegatePubKey!=b2.delegatePubKey) {
      // Blocks by different delegates are never contradicting
      return false;
   } else if(b1.blockID==b2.blockID) {
      // No contradiction, as block headers are the same
      return false;
   } else if (b1.maxHeightPrevoted==b2.maxHeightPrevoted &&  b1.height>=b2.height) {
      // Violation of the fork choice rule as delegate moved to different chain
      // without strictly larger maxHeightPrevoted or larger height as justification.
      // This in particular happens, if a delegate is double forging.
      return true;
   } else if(b1.height>b2.maxHeightPreviouslyForged) {
      // Violates disjointness condition
      return true;
   } else if(b1.maxHeightPrevoted>b2.maxHeightPrevoted) {
      // Violates that delegate chooses branch with largest maxHeightPrevoted
      return true;
   } else {
      // No contradiction between block headers
      return false;
   }
}
----
(Source: {url_github_lip14}[LIP 14^])

=== Reporting a delegate

> A xref:{url_protocol_transactions_pom}[delegate misbehavior report] transaction can be issued by anyone in the network who observes a violation of the Lisk BFT consensus algorithm, as explained in <<discover>>.

Reviewer 3: ```suggestion
A xref:{url_protocol_transactions_pom}[delegate misbehavior report] transaction can be issued by anyone in the network who observes a violation of the Lisk-BFT consensus algorithm, as explained in <<discover>>.
```

            totalFee: '0',
            version: 2
        },
        header2: {
            blockSignature: '31ccf4ce1a3a224a2a32c3f4bdc6fad0ddb8feb45b05b7d411eee1a608f9d91284d09c727bba173c882d5dc90cb951c5affc10462d650031a627e00d919cbf08',
            generatorPublicKey: '7a7f24c061db6a92320ba14323f814c20dbcc811a931ead3ca63c75a4de1b643',
            height: 8933,
            maxHeightPreviouslyForged: 8788,
            maxHeightPrevoted: 8868,
            numberOfTransactions: 0,
            payloadHash: 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855',
            payloadLength: 0,
            previousBlockId: '9326981395427095175',
            reward: '500000000',
            seedReveal: 'abe2a66d7a35fd7b580e977d9f7911ae',
            timestamp: 122329567,
            totalAmount: '0',
            totalFee: '0',
            version: 2
        }
    }
});
----

[[punishment_consequences]]
=== Consequences for punished delegates and their voters

If a valid delegate misbehavior report is posted to the network, the respective delegate will face the following consequences:

* Setting the delegate weight to 0 for the next 780,000 blocks (approximately 3 months).

> * The unlocking period for self-votes is increased from 260,000 blocks to 780,000 blocks (from approximately 1 month to 3 months), for the respective delegate.

Reviewer 3: ```suggestion
* The unlocking period for self-votes is increased from 260,000 blocks to 780,000 blocks (from approximately 1 month to 3 months).
```

Reviewer 3: We already say "for the respective delegate" before the bullet list.

> Voters of the respective delegate will receive the following punishment:

Reviewer 3: ```suggestion
Any account voting for the respective delegate will receive the following punishment:
```

This means that for a given height, eventually all Lisk nodes agree on the same block.
This is particularly important in the case whereby there are different valid blocks for the same height, which can occur due to network delays or delegates forging multiple blocks in their designated time slot.

In most cases, the BFT consensus algorithm implemented in the Lisk SDK takes care of these points and resolves all conflicts automatically, following the rules defined in the xref:{url_protocol_bft}[BFT] section of the Lisk Protocol.

Additionally, delegates can be punished by anyone in the network, if they forge contradicting blocks, i.e., two blocks with consensus votes that violate the Lisk BFT (see the xref:{url_protocol_punishment}[Punishment of Lisk-BFT protocol violations] section of the Lisk Protocol).
This can happen for example if the following actions are performed:

* ... Double forging, i.e. having forging activated on multiple nodes for the same delegate at the same time.
* ... Violating the xref:{url_protocol_fork_choice}[fork choice rule].
* ... xref:{url_guides_forging_safely}[Enabling forging on a new node], without porting over the `forger_info` data.

> === Delegate misbehaviour report

Reviewer: ```suggestion
=== Delegate misbehavior report
```

});
----

.A token unlock transaction can contain multiple unlock objects
[TIP]
====
This allows an account to submit multiple delegate votes and recover those tokens with a single unlock transaction.
Of course, all tokens must have been in the unlocking state for at least 2000 blocks for the unlock to be valid.
====

== BFT

The Byzantine Fault Tolerance (BFT) algorithm ensures that the network can reach consensus about the current state of the blockchain.
This means that for a given height, eventually all Lisk nodes agree on the same block.

> This is particularly important in the case whereby there are different valid blocks for the same height, which can occur due to network delays or delegates forging multiple blocks in their designated time slot.

Reviewer 3: ```suggestion
This is particularly important in the case where there are different valid blocks for the same height, which can occur due to network delays or delegates forging multiple blocks in their designated time slot.
```

In most cases, the BFT consensus algorithm implemented in the Lisk SDK takes care of these points and resolves all conflicts automatically, following the rules defined in the xref:{url_protocol_bft}[BFT] section of the Lisk Protocol.

> Additionally, delegates can be punished by anyone in the network, if they forge contradicting blocks, i.e., two blocks with consensus votes that violate the Lisk BFT (see the xref:{url_protocol_punishment}[protocol documentation] for details).

Reviewer 3: ```suggestion
Additionally, delegates can be punished by anyone in the network, if they forge contradicting blocks, i.e., two blocks with consensus votes that violate the Lisk-BFT protocol (see the xref:{url_protocol_punishment}[protocol documentation] for details).
```

> In order to avoid that a delegate is punished, a node operator has to take a lot of care when enabling forging on a node.

Reviewer: ```suggestion
In order to avoid that a delegate is punished, a node operator has to take great care when enabling forging on a node.
```

This means that they have to ensure the following:

* Never activate forging on more than one node.
Otherwise, the delegate may be punished due to double-forge, i.e., producing two contradicting blocks for the same time slot.
* Never activate forging without porting over the `forger_info` data (see xref:{url_guides_forging_safely}[safely enabling forging on a new node]).
Otherwise, the `maxHeightPreviouslyForged` property of the forged blocks may be outdated, which can cause the delegate to forge contradicting blocks.

[[discover]]
=== How to discover misbehavior of a delegate

Any misbehavior of a delegate is indicated by two contradicting block headers signed by the delegate.
A misbehavior can therefore be reported by providing the two contradicting block headers.

> The code below checks the conditions for two block headers to be contradicting.

Reviewer: ```suggestion
The code below checks the conditions in order to see if two contradicting block headers exist.
```

Author: ```suggestion
The code below checks if two contradicting block headers exist.
```

      b1=blockHeader2;
      b2=blockHeader1;
   }

   // The order of cases is essential here
   if(b1.delegatePubKey!=b2.delegatePubKey) {
      // Blocks by different delegates are never contradicting
      return false;
   } else if(b1.blockID==b2.blockID) {
      // No contradiction, as block headers are the same
      return false;
   } else if (b1.maxHeightPrevoted==b2.maxHeightPrevoted &&  b1.height>=b2.height) {
      // Violation of the fork choice rule as delegate moved to different chain
      // without strictly larger maxHeightPrevoted or larger height as justification.
      // This in particular happens, if a delegate is double forging.
      return true;
   } else if(b1.height>b2.maxHeightPreviouslyForged) {
      // Violates disjointness condition
      return true;
   } else if(b1.maxHeightPrevoted>b2.maxHeightPrevoted) {
      // Violates that delegate chooses branch with largest maxHeightPrevoted
      return true;
   } else {
      // No contradiction between block headers
      return false;
   }
}
----
(Source: {url_github_lip14}[LIP 14^])

=== Reporting a delegate

> A xref:{url_protocol_transactions_pom}[delegate misbehaviour report] transaction can be issued by anyone in the network who observes a violation of the Lisk BFT consensus algorithm, like explained in <<discover>>.

Reviewer: ```suggestion
A xref:{url_protocol_transactions_pom}[delegate misbehavior report] transaction can be issued by anyone in the network who observes a violation of the Lisk BFT consensus algorithm, as explained in <<discover>>.
```
